File: 2070-most-beautiful-item-for-each-query/2070-most-beautiful-item-for-each-query.java
class Solution {
    public int[] maximumBeauty(int[][] items, int[] queries) {
        
        Map<Integer,Integer> map = new TreeMap<>();  //For Sorting and removing duplicates
        for(int[] a:items)
        {
            if(map.get(a[0])==null)
                map.put(a[0],a[1]);
            else if(map.get(a[0])<a[1])
                map.put(a[0],a[1]);
        }
        
        Map<Integer,Integer> map2 = new HashMap<>();  //For storing beautiful items for keySet
        int max=0;
        
        for(var entry:map.entrySet())
        {
            if(entry.getValue()>max)
            {
                map2.put(entry.getKey(),entry.getValue());
                max=entry.getValue();
            }
            else
                map2.put(entry.getKey(),max);
        }
        
        int i=0;
        int[] res = new int[queries.length];
     ArrayList<Integer> list = new ArrayList<>(map.keySet()); //Sorted KeySet for applying Binary Search
        for(int x:queries)
        {
            int index=0;
            if(map2.get(x)==null)
            {
                index = Collections.binarySearch(list,x)*-1-2;
                res[i++] = index<0? 0:map2.get(list.get(index)); 
            }
            else
                res[i++]=map2.get(x);
        }
        return res;
    }
}

File: 2070-most-beautiful-item-for-each-query/2070-most-beautiful-item-for-each-query.py
class Solution:
    def maximumBeauty(self, items: List[List[int]], queries: List[int]) -> List[int]:
        items.sort()
        n = len(items)
        res = []
        for i in range(1, n):
            items[i][1] = max(items[i][1], items[i - 1][1])

        for q in queries:
            l, r = 0, n - 1
            while l <= r:
                m = (l + r) >> 1
                p, b = items[m]
                if p > q: r = m - 1
                else: l = m + 1
            res.append(items[r][1] if r >= 0 else 0)
        return res